src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Protocol.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.ProjectSystem;
+
+/// <summary>
+/// Publishes project data (including TagHelper info) discovered OOB to the server via LSP notification
+/// instead of old method of writing a project configuration bin file
+/// </summary>
+[Shared]
+[Export(typeof(RazorProjectInfoEndpointPublisher))]
+internal partial class RazorProjectInfoEndpointPublisher : IDisposable
+{
+    private readonly LSPRequestInvoker _requestInvoker;
+    private readonly IProjectSnapshotManager _projectManager;
+
+    private readonly AsyncBatchingWorkQueue<(IProjectSnapshot Project, bool Removal)> _workQueue;
+    private readonly CancellationTokenSource _disposeTokenSource;
+    private bool _active;
+
+    [ImportingConstructor]
+    public RazorProjectInfoEndpointPublisher(
+        LSPRequestInvoker requestInvoker,
+        IProjectSnapshotManager projectManager
+    )
+    {
+        _requestInvoker = requestInvoker;
+        _projectManager = projectManager;
+
+        _disposeTokenSource = new ();
+        _workQueue = new(
+            EnqueueDelay,
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    // internal for testing
+    // Delay between publishes to prevent bursts of changes yet still be responsive to changes.
+    internal TimeSpan EnqueueDelay { get; set; } = TimeSpan.FromMilliseconds(250);
```

Comment:
Tests can't actually use this since it is consumed in the constructor. See my similar comment in `ProjectConfiguratonStateManager.cs` for a potential solution.